h2quic/response_writer.go

```diff
@@ -86,12 +98,129 @@ func (w *responseWriter) Flush() {}
+		}
+		u.Scheme = wantScheme
+		u.Host = w.requestHost
+	} else {
+		if u.Scheme != wantScheme {
+			return fmt.Errorf("cannot push URL with scheme %q from request with scheme %q", u.Scheme, wantScheme)
+		}
+		if u.Host == "" {
+			return errors.New("URL must have a host")
+		}
+	}
+	for k := range opts.Header {
+		if strings.HasPrefix(k, ":") {
+			return fmt.Errorf("promised request headers cannot include pseudo header %q", k)
+		}
+		switch strings.ToLower(k) {
+		case "content-length", "content-encoding", "trailer", "te", "expect", "host":
+			return fmt.Errorf("promised request headers cannot include %q", k)
+		}
+	}
+	if err = checkValidHTTP2RequestHeaders(opts.Header); err != nil {
+		return err
+	}
+	authority := u.Host
+	path := u.RequestURI()
+	contentLengthStr := "0" // a request does not have content
+
+	// Construct HTTP headers for request in push promise
+	var headers bytes.Buffer
+	enc := hpack.NewEncoder(&headers)
+	enc.WriteField(hpack.HeaderField{Name: ":method", Value: opts.Method})
+	enc.WriteField(hpack.HeaderField{Name: ":path", Value: path})
+	enc.WriteField(hpack.HeaderField{Name: ":authority", Value: authority})
+	enc.WriteField(hpack.HeaderField{Name: ":scheme", Value: u.Scheme})
+	for k, v := range opts.Header {
+		for index := range v {
+			enc.WriteField(hpack.HeaderField{Name: strings.ToLower(k), Value: v[index]})
+		}
+	}
+
+	// Get new data stream
+	newDataStream, err := w.session.OpenStream()
+	if err != nil {
+		if err == qerr.TooManyOpenStreams {
+			return http2.ErrPushLimitReached
+		}
+		return err
+	}
+	newDataStreamID := newDataStream.StreamID()
+	if newDataStreamID%2 != 0 {
+		return fmt.Errorf("Cannot push on client side stream")
+	}
+
+	// Write push promise header
+	utils.Debugf("Sending PUSH_PROMISE for target '%s', promised on stream %d", target, newDataStreamID)
+	headerFramer := http2.NewFramer(w.headerStream, nil)
+	w.headerStreamMutex.Lock()
+	err = headerFramer.WritePushPromise(http2.PushPromiseParam{
+		StreamID:      uint32(w.dataStreamID),
+		PromiseID:     uint32(newDataStreamID),
+		BlockFragment: headers.Bytes(),
+		EndHeaders:    true,
+	})
+	w.headerStreamMutex.Unlock() // Do not defer as we will first call ServeHTTP(), which will need the mutex
+	if err != nil {
+		return err
+	}
+
+	// golang net/http2 constructs a 'fake' http2 request and feeds it to the serve() loop to let it be processed as a normal request.
+	// But we will, for now, just serveHTTP() it here:
+	pushRequest, err := requestFromHTTPHeader(opts.Header, path, authority, opts.Method, contentLengthStr)
+	if err != nil {
+		return err
+	}
+	pushRequest.RemoteAddr = w.session.RemoteAddr().String()
+	reqBody := newRequestBody(newDataStream)
+	pushRequest.Body = reqBody
+	pushRequestResponseWriter := newResponseWriter(w.headerStream, w.headerStreamMutex, newDataStream, newDataStreamID, w.session, w.handlerFunc, w.requestHost)
+	serveHTTP(w.handlerFunc, pushRequestResponseWriter, pushRequest, true, reqBody)
```

Comment:
As far as I know, the stream has to be closed by the server when the push resource is sent. 

As for Google QUIC, I think the server also closes this stream immediately after sending the data:
```
[0705/020215.411669:VERBOSE1:quic_simple_server_stream.cc(246)] Stream 2 writing body (fin = 1) with size: 1559
[0705/020215.411729:VERBOSE1:quic_framer.cc(650)] Server: Appending header: { connection_id: 17336184730930128488, connection_id_length: 0, packet_number_length: 1, reset_flag: 0, version_flag: 0, packet_number: 8 }
[0705/020215.411779:VERBOSE1:quic_packet_creator.cc(393)] Server: Adding frame: { stream_id: 2, fin: 0, offset: 0, length: 1334 }
[0705/020215.411862:VERBOSE1:quic_connection.cc(1516)] Server: Sending packet 8 : data bearing , encryption level: ENCRYPTION_FORWARD_SECURE, encrypted length:1350
[0705/020215.412002:VERBOSE1:quic_connection.cc(1589)] Server: time we began writing last sent packet: 169139771276
[0705/020215.412055:VERBOSE1:quic_framer.cc(650)] Server: Appending header: { connection_id: 17336184730930128488, connection_id_length: 0, packet_number_length: 1, reset_flag: 0, version_flag: 0, packet_number: 9 }
[0705/020215.412102:VERBOSE1:quic_packet_creator.cc(393)] Server: Adding frame: { stream_id: 2, fin: 1, offset: 1334, length: 225 }
[0705/020215.412167:VERBOSE1:quic_connection.cc(1516)] Server: Sending packet 9 : data bearing , encryption level: ENCRYPTION_FORWARD_SECURE, encrypted length:243
[0705/020215.412277:VERBOSE1:quic_connection.cc(1589)] Server: time we began writing last sent packet: 169139771581
[0705/020215.412330:VERBOSE1:quic_flow_controller.cc(96)] Server: Stream 2 sent 1559 bytes.
[0705/020215.412368:VERBOSE1:quic_flow_controller.cc(96)] Server: Stream 0 sent 1559 bytes.
[0705/020215.412421:VERBOSE1:quic_stream.cc(387)] Server: Done writing to stream 2
[0705/020215.412458:VERBOSE1:quic_stream.cc(391)] Server: Closing stream 2
[0705/020215.412495:VERBOSE1:quic_session.cc(355)] Server: Closing stream 2
```